### PR TITLE
added function put_along_axis in indexing_like_operations

### PR DESCRIPTION
**Pull request by PM-23**
****Add Indexing routines for Numpy Frontend #1536**

**put_along_axis #13874****

### CODE REVIEW

**Comment by MuhammadSaeedBatikh** [comment]
Hi @PM-23, please let me know when you think it's ready for review. Thanks.

**Review by PM-23 — COMMENTED** [review]
Hello @MuhammadSaeedBatikh  I've resolved the conflicts and updated the code as well for the function **put_along_axis()** Please can you check it now? Thanks!

**Comment by MuhammadSaeedBatikh** [comment]
Hi @PM-23, thanks for the PR.
There's still a missing test function. You should implement `test_numpy_put_along_axis` function with the appropriate parameters in `ivy_tests/test/ivy/test_frontends/test_numpy/test_indexing_routines/test_indexing_like_operations.py` .

**Comment by PM-23** [comment]
Hello @MuhammadSaeedBatikh thanks for informing. Well I have one doubt as I would implement the function **test_numpy_put_along_axis** in **test_indexing_like_operations.py** file but as the **put_along_axis** function has one extra parameter '_values_' and there's no such kind of parameter in the helper function **helpers.array_indices_axis()**. So do I need to create another helper function for the put_along_axis to implement. So can you please let me know. Thanks!

**Comment by MuhammadSaeedBatikh** [comment]
> Hello @MuhammadSaeedBatikh thanks for informing. Well I have one doubt as I would implement the function **test_numpy_put_along_axis** in **test_indexing_like_operations.py** file but as the **put_along_axis** function has one extra parameter '_values_' and there's no such kind of parameter in the helper function **helpers.array_indices_axis()**. So do I need to create another helper function for the put_along_axis to implement. So can you please let me know. Thanks!

Yes, but I think you might want to try a simpler option first before implementing a new function. You can create a simple strategy for the values parameters and still use `helpers.array_indices_axis()`  for the rest.

**Comment by PM-23** [comment]
> > Hello @MuhammadSaeedBatikh thanks for informing. Well I have one doubt as I would implement the function **test_numpy_put_along_axis** in **test_indexing_like_operations.py** file but as the **put_along_axis** function has one extra parameter '_values_' and there's no such kind of parameter in the helper function **helpers.array_indices_axis()**. So do I need to create another helper function for the put_along_axis to implement. So can you please let me know. Thanks!
> 
> Yes, but I think you might want to try a simpler option first before implementing a new function. You can create a simple strategy for the values parameters and still use `helpers.array_indices_axis()` for the rest.

Okay @MuhammadSaeedBatikh  I updated as you mentioned. Please check and let me know. Thanks.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 1]
```diff
@@ -780,7 +784,20 @@ def array_indices_axis(
     indices = indices[0]
     if disable_random_axis:
         return [x_dtype, indices_dtype], x, indices
-    return [x_dtype, indices_dtype], x, indices, axis, batch_dims
+
+    if values is not None:
```
Comment (MuhammadSaeedBatikh) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:788]: This function is used by other tests. In order not to break them you should maintain the original output on `values = None`:
 ```
if values is None:
    return [x_dtype, indices_dtype], x, indices, axis, batch_dims
else:
      ....
```

File: ivy/functional/ivy/experimental/manipulation.py [hunk 2]
```diff
@@ -1613,7 +1613,87 @@ def expand(
     ret
         Output Array
     """
-    return ivy.current_backend(x).expand(x, shape, out=out, copy=copy)
+    return ivy.current_backend(x).expand(x, shape, out=out)
```
Comment (MuhammadSaeedBatikh) [inline ivy/functional/ivy/experimental/manipulation.py:1616]: You shouldn't change the behavior of other functions. You can just pass `copy=None` when calling `expand`  from `put_along_axis`

**Comment by MuhammadSaeedBatikh** [comment]
Hi @PM-23, 
I left some comments on your new modifications. Also, tests are not running for `put_along_axis` function because of an import error. Replace `fn_tree="ivy.put_along_axis"` to `fn_tree="numpy.put_along_axis"`

File: ivy/functional/ivy/experimental/manipulation.py (hunk 2 above)
Comment (PM-23) [reply ivy/functional/ivy/experimental/manipulation.py]: Okay, I apologise that was by mistake while solving merge conflicts.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py (hunk 1 above)
Comment (PM-23) [reply ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py]: Okay sure! Thanks.

**Review by PM-23 — COMMENTED** [review]
Hi @MuhammadSaeedBatikh thanks. I updated with required changes. Thank you!

**Comment by MuhammadSaeedBatikh** [comment]
Hi @PM-23,
I think there's still an unpacking issue with the helper function.
```
test_ivy_put_along_axis
dtypes, x, indices, axis, values, _ = dtype_x_indices_axis
  E ValueError: not enough values to unpack (expected 6, got 5)
   Falsifying example: test_ivy_put_along_axis(
       dtype_x_indices_axis=(['int16', 'int32'],
        array([0], dtype=int16),
        array([0], dtype=int32),
        0,
        0),
       test_flags=FrontendFunctionTestFlags(
           num_positional_args=0,
          with_out=False,
           inplace=False,
           as_variable=[False],
           native_arrays=[False],
           generate_frontend_arrays=False,
       ),
       fn_tree='ivy.functional.frontends.numpy.put_along_axis',
       frontend='numpy',
       on_device='cpu',
)
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 3]
```diff
@@ -780,7 +784,21 @@ def array_indices_axis(
     indices = indices[0]
     if disable_random_axis:
         return [x_dtype, indices_dtype], x, indices
-    return [x_dtype, indices_dtype], x, indices, axis, batch_dims
+
+    if values is None:
+        return [x_dtype, indices_dtype], x, indices, axis, batch_dims
+    else:
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:790]: Hello @PM-23, I'd suggest moving this logic to a composite strategy in the test file, instead of having to change the strategy itself, we shouldn't be changing this one here for a single test.

**Review by CatB1t — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py (hunk 3 above)
Comment (PM-23) [reply ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:790]: Hi @CatB1t so could you please let me know I'm confused so to make it clear I need to update the code in only test file and not in helper file? But as it needs to have **values** parameter that needs to be returned when put_along_axis function is used. So can you provide me some more information for your suggestion. Thanks!

**Comment by PM-23** [comment]
Hi @MuhammadSaeedBatikh I updated the code and changed test put_along_axis function can you please check and let me know. Thanks!

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py (hunk 3 above)
Comment (PM-23) [reply ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:794]: Hi @CatB1t could you please have a look again?

Comment (CatB1t) [reply ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:794]: @PM-23, in that case we should try to define our composite strategy with a different behavior in the test file of the function using the `@st.composite` decorator, You can use the strategies defined here: https://unify.ai/docs/ivy/docs/helpers/ivy_tests.test_ivy.helpers.hypothesis_helpers.html to help you build your custom strategy.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 4]
```diff
@@ -781,7 +788,21 @@ def array_indices_axis(
     indices = indices[0]
     if disable_random_axis:
         return [x_dtype, indices_dtype], x, indices
-    return [x_dtype, indices_dtype], x, indices, axis, batch_dims
+
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:791]: Please, revert the changes to the strategy, we should ideally create a custom one instead of modifying the array strategy to support an edge case.

**Review by CatB1t — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py (hunk 4 above)
Comment (PM-23) [reply ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:784]: Okay @CatB1t , I've reverted the changes and created a custom one please check it! Thanks

**Comment by PM-23** [comment]
I fixed a keyword argument error with **ivy.expand**, and some **linting errors**.

**Comment by MuhammadSaeedBatikh** [comment]
Hi @PM-23, thanks a lot for your work. 
The remaining failing tests do not seem related to the task. Everything lgtm. 

> I fixed a keyword argument error with **ivy.expand**, and some **linting errors**.
